source/common/router/router.h

```diff
@@ -371,32 +394,41 @@ class Filter : Logger::Loggable<Logger::Id::router>,
                                          Upstream::ResourcePriority priority) PURE;
   Http::ConnectionPool::Instance* getConnPool();
   void maybeDoShadowing();
-  bool maybeRetryReset(Http::StreamResetReason reset_reason);
-  void onPerTryTimeout();
+  bool maybeRetryReset(Http::StreamResetReason reset_reason, UpstreamRequest* upstream_request);
+  uint32_t numRequestsAwaitingHeaders();
+  void onGlobalTimeout();
+  void onPerTryTimeout(UpstreamRequest* upstream_request);
   void onRequestComplete();
   void onResponseTimeout();
-  void onUpstream100ContinueHeaders(Http::HeaderMapPtr&& headers);
+  void onUpstream100ContinueHeaders(Http::HeaderMapPtr&& headers,
+                                    UpstreamRequest* upstream_request);
   // Handle an upstream request aborted due to a local timeout.
+  void onSoftPerTryTimeout();
+  void onSoftPerTryTimeout(UpstreamRequest* upstream_request);
   void onUpstreamTimeoutAbort(StreamInfo::ResponseFlag response_flag);
   // Handle an "aborted" upstream request, meaning we didn't see response
   // headers (e.g. due to a reset). Handles recording stats and responding
   // downstream if appropriate.
   void onUpstreamAbort(Http::Code code, StreamInfo::ResponseFlag response_flag,
                        absl::string_view body, bool dropped);
-  void onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& headers, bool end_stream);
-  void onUpstreamData(Buffer::Instance& data, bool end_stream);
-  void onUpstreamTrailers(Http::HeaderMapPtr&& trailers);
+  void onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& headers,
+                         UpstreamRequest* upstream_request, bool end_stream);
+  void onUpstreamData(Buffer::Instance& data, UpstreamRequest* upstream_request, bool end_stream);
+  void onUpstreamTrailers(Http::HeaderMapPtr&& trailers, UpstreamRequest* upstream_request);
   void onUpstreamMetadata(Http::MetadataMapPtr&& metadata_map);
```

Comment:
not sure if anything needs to change with this function. right now all upstream metadata gets proxied downstream